desc/optimize/_constraint_wrappers.py

```diff
@@ -569,32 +493,76 @@ def build(self, eq=None, use_jit=None, verbose=1):  # noqa: C901
         for constraint in self._linear_constraints:
             constraint.build(self._eq, verbose=verbose)
 
+        self._objectives = combine_args(self._objective, self._constraint)
+        self._set_things(self._all_things)
+
         self._dim_f = self._objective.dim_f
         if self._dim_f == 1:
             self._scalar = True
         else:
             self._scalar = False
 
-        self.set_args()
-        self._objective.set_args(*self._full_args)
-        self._constraint.set_args(*self._full_args)
+        self._set_state_vector()
         # history and caching
-        self._x_old = np.zeros((self._dim_x,))
-        for arg in self.args:
-            self._x_old[self.x_idx[arg]] = getattr(eq, arg)
-
+        self._x_old = self.x(eq)
         self._allx = [self._x_old]
         self._allxopt = [self._objective.x(eq)]
         self._allxeq = [self._constraint.x(eq)]
-        self.history = {}
-        for arg in arg_order:
-            self.history[arg] = [np.asarray(getattr(self._eq, arg)).copy()]
+        self.history = [[self._eq.params_dict]]
 
         self._built = True
         timer.stop("Proximal projection build")
         if verbose > 1:
             timer.disp("Proximal projection build")
 
+    def unpack_state(self, x, per_objective=True):
+        """Unpack the state vector into its components.
+
+        Parameters
+        ----------
+        x : ndarray
+            State vector.
+        per_objective : bool
+            Whether to return param dicts for each objective (default) or for each
+            unique optimizable thing.
+
+        Returns
+        -------
+        params : dict
+            Parameter dictionary for equilibrium, with just external degrees of freedom
+            visible to the optimizer.
+
+        """
+        if not self.built:
+            raise RuntimeError("ObjectiveFunction must be built first.")
+
+        x = jnp.atleast_1d(x)
+        if x.size != self.dim_x:
+            raise ValueError(
+                "Input vector dimension is invalid, expected "
+                + f"{self.dim_x} got {x.size}."
+            )
+
+        xs_splits = np.cumsum([self._eq.dimensions[arg] for arg in self._args])
+        # kludge for now if things is only equilibrium
+        params = [{arg: xs for arg, xs in zip(self._args, jnp.split(x, xs_splits))}]
+        if per_objective:
+            params = self._unflatten(params)
+        return params
+
+    def x(self, *things):
+        """Return the full state vector from the Optimizable objects things."""
+        # TODO: also check resolution etc?
+        assert [type(t1) == type(t2) for t1, t2 in zip(things, self.things)]
```

Comment:
Why do we have to check against `self.things`, just as a safety measure? And why not have the input `*things` default to `self.things`?

Reply:
yeah its a safety check. Ideally we'd also check specific resolution etc but that gets messy. I'll make it default to self.things (same as for the regular ObjectiveFunction)